File: app/src/main/java/com/example/task2/ui/benchmark/BenchmarksViewModel.java
package com.example.task2.ui.benchmark;

import android.util.Log;

import androidx.core.util.Pair;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.task2.R;
import com.example.task2.models.Benchmark;
import com.example.task2.models.CellOperation;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class BenchmarksViewModel extends ViewModel {
    private final Benchmark benchmark;
    private final MutableLiveData<List<CellOperation>> cellOperationsLiveData = new MutableLiveData<>();
    private final MutableLiveData<Boolean> allTasksCompletedLiveData = new MutableLiveData<>();
    private final MutableLiveData<Integer> validNumberLiveData = new MutableLiveData<>();
    private Disposable disposable = Disposable.disposed();

    public BenchmarksViewModel(Benchmark benchmark) {
        this.benchmark = benchmark;
    }

    public LiveData<List<CellOperation>> getCellOperationsLiveData() {
        return cellOperationsLiveData;
    }

    public LiveData<Boolean> getAllTasksCompletedLiveData() {
        return allTasksCompletedLiveData;
    }

    public LiveData<Integer> getValidNumberLiveData() {
        return validNumberLiveData;
    }

    public void onButtonClicked(String input) {
        if (disposable != null && !disposable.isDisposed()) {
            stopBenchmark();
        } else {
            final int number = Integer.parseInt(input);
            runBenchmark(number);
        }
    }

    public void validateNumber(String input) {
        int number;
        Integer errorMessage = null;

        try {
            number = Integer.parseInt(input);
            if (number < 1) {
                errorMessage = R.string.error_count;
            }
        } catch (NumberFormatException e) {
            errorMessage = R.string.error_valid;
        }

        validNumberLiveData.setValue(errorMessage);
    }

    public void onCreate() {
        cellOperationsLiveData.setValue(benchmark.createItemsList(false));
    }

    public int getNumberOfColumns() {
        return benchmark.getNumberOfColumns();
    }

    public void runBenchmark(int number) {
        final List<CellOperation> operations = benchmark.createItemsList(true);
        cellOperationsLiveData.postValue(new ArrayList<>(operations));

        disposable = Observable.fromIterable(operations)
                .flatMap(cell -> Observable.fromCallable(() -> {
                    final long operationTime = benchmark.measureTime(cell, number);
                    return new Pair<>(operations.indexOf(cell), cell.withTime(operationTime));
                }))
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .doFinally(() -> allTasksCompletedLiveData.setValue(true))
                .subscribe(pair -> {
                    int position = pair.first;
                    CellOperation updatedCell = pair.second;
                    operations.set(position, updatedCell);
                    cellOperationsLiveData.setValue(new ArrayList<>(operations));
                }, throwable -> Log.e("LOGG:", "Error: " + throwable.getMessage()));
    }

    public void stopBenchmark() {
        disposable.dispose();

        final List<CellOperation> list = new ArrayList<>(Objects.requireNonNull(cellOperationsLiveData.getValue()));
        for (int i = 0; i < list.size(); i++) {
            if (list.get(i).isRunning) {
                list.set(i, list.get(i).withIsRunning(false));
            }
        }
        cellOperationsLiveData.setValue(list);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        if (disposable != null && !disposable.isDisposed()) {
            disposable.dispose();
        }
    }
}

File: app/src/main/java/com/example/task2/ui/benchmark/BenchmarksViewModel.kt
package com.example.task2.ui.benchmark

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.task2.BenchmarksApp
import com.example.task2.R
import com.example.task2.models.benchmarks.Benchmark
import com.example.task2.models.benchmarks.CellOperation
import com.example.task2.models.room.AppDatabase
import com.example.task2.models.room.OperationResult
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class BenchmarksViewModel(private val benchmark: Benchmark) : ViewModel() {

	private val operationResultDao = AppDatabase.getInstance().operationResultDao()
	private val cellOperationsLiveData = MutableLiveData<List<CellOperation>>(emptyList())
	private val allTasksCompletedLiveData = MutableLiveData(true)
	private val validNumberLiveData = MutableLiveData<Int?>()

	private var benchmarkJob: Job? = null

	fun getCellOperationsLiveData(): LiveData<List<CellOperation>> = cellOperationsLiveData
	fun getAllTasksCompletedLiveData(): LiveData<Boolean> = allTasksCompletedLiveData
	fun getValidNumberLiveData(): LiveData<Int?> = validNumberLiveData
	fun getNumberOfColumns(): Int = benchmark.getNumberOfColumns()

	fun onButtonClicked(input: String) {
		val number = input.toIntOrNull()
		println("Job running ${benchmarkJob?.isActive}, completed  ${benchmarkJob?.isCompleted}")
		when {
			benchmarkJob?.isActive == true -> stopBenchmark()
			number != null -> runBenchmark(number)
		}
	}

	fun validateNumber(input: String) {
		val number = input.toIntOrNull()
		val errorMessage = when {
			number == null -> R.string.error_valid
			number < 1 -> R.string.error_count
			else -> null
		}
		validNumberLiveData.value = errorMessage
	}

	fun onCreate() {
		cellOperationsLiveData.value = benchmark.createItemsList(false)
	}

	private fun addResultToDatabase(result: List<CellOperation>, input: Int) {
		viewModelScope.launch(Dispatchers.IO) {
			for (cell in result) {
				val appContext = BenchmarksApp.instance
				val action = appContext?.getString(cell.action)
				val type = appContext?.getString(cell.type)
				if (action != null && type != null) {
					val operationResult = OperationResult(
						action = action,
						type = type,
						time = cell.time,
						input = input
					)
					try {
						operationResultDao.insertResult(operationResult)
					} catch (e: Exception) {
						e.printStackTrace()
					}
				}
			}
		}
	}

	suspend fun getLast21Results(): List<OperationResult> {
		return withContext(viewModelScope.coroutineContext + Dispatchers.IO) {
			operationResultDao.getLast21Results()
		}
	}

	private fun runBenchmark(number: Int) {
		val handler = CoroutineExceptionHandler { _, exception ->
			exception.printStackTrace()
		}

		val operations = benchmark.createItemsList(true).toMutableList()
		cellOperationsLiveData.value = ArrayList(operations)

		benchmarkJob = viewModelScope.launch(Dispatchers.IO + handler) {
			operations.mapIndexed { index, cell ->
				async {
					val operationTime = benchmark.measureTime(cell, number)
					val updatedCell = cell.withTime(operationTime)

					withContext(Dispatchers.Main) {
						operations[index] = updatedCell
						cellOperationsLiveData.value = ArrayList(operations)
					}
				}
			}.awaitAll()
			addResultToDatabase(operations, number)
			withContext(Dispatchers.Main) {
				allTasksCompletedLiveData.value = true
			}
		}
	}

	private fun stopBenchmark() {
		benchmarkJob?.cancel()
		cellOperationsLiveData.value = cellOperationsLiveData.value?.map { it.withIsRunning(false) }
	}
}
